src/Scripting/Core/Hosting/Resolvers/RuntimeMetadataReferenceResolver.cs

```diff
@@ -106,28 +131,90 @@ public override ImmutableArray<PortableExecutableReference> ResolveReference(str
                     if (resolvedPath != null)
                     {
                         return ImmutableArray.Create(_fileReferenceProvider(resolvedPath, properties));
                     }
                 }
             }
-            else if (GacFileResolver != null)
+            else
             {
-                var path = GacFileResolver.Resolve(reference);
-                if (path != null)
+                if (GacFileResolver != null)
+                {
+                    string path = GacFileResolver.Resolve(reference);
+                    if (path != null)
+                    {
+                        return ImmutableArray.Create(_fileReferenceProvider(path, properties));
+                    }
+                }
+
+                if (_useCoreResolver && AssemblyIdentity.TryParseDisplayName(reference, out var identity, out var identityParts))
                 {
-                    return ImmutableArray.Create(_fileReferenceProvider(path, properties));
+                    var result = ResolveTrustedPlatformAssemblyCore(identity.Name, properties);
+                    if (result != null)
+                    {
+                        return ImmutableArray.Create(result);
+                    }
                 }
             }
 
             return ImmutableArray<PortableExecutableReference>.Empty;
         }
 
+        private PortableExecutableReference ResolveTrustedPlatformAssemblyCore(string name, MetadataReferenceProperties properties)
+        {
+            if (_lazyTrustedPlatformAssemblies == null)
+            {
+                _lazyTrustedPlatformAssemblies = GetTrustedPlatformAssemblyMap();
+            }
+
+            if (_lazyTrustedPlatformAssemblies.TryGetValue(name, out string path) && File.Exists(path))
+            {
+                return MetadataReference.CreateFromFile(path, properties);
+            }
+
+            return null;
+        }
+
+        private static ImmutableDictionary<string, string> GetTrustedPlatformAssemblyMap()
+        {
+            var set = ImmutableDictionary.CreateBuilder<string, string>(StringComparer.OrdinalIgnoreCase);
+
+            if (CoreClrShim.AppContext.GetData?.Invoke("TRUSTED_PLATFORM_ASSEMBLIES") is string paths)
+            {
+                foreach (var path in paths.Split(Path.PathSeparator))
+                {
+                    try
+                    {
+                        if (Path.GetExtension(path) == ".dll")
+                        {
+                            string fileName = Path.GetFileNameWithoutExtension(path);
+                            if (fileName.EndsWith(".ni", StringComparison.OrdinalIgnoreCase))
+                            {
+                                fileName = fileName.Substring(0, fileName.Length - ".ni".Length);
+                            }
+
+                            // last one wins:
+                            set[fileName] = path;
+                        }
+                    }
```

Comment:
The scope of this catch is big enough to trigger my concern that a bug could be eaten. Can you reduce the scope of tighten the exception types caught?